Drop some TSan suppressions

hebasto opened the pull request:
It seems possible now to drop some TSan suppressions.

hebasto committed:
Update TSan suppressions

practicalswift commented:
Concept ACK

Great to see `deadlock:UpdateTip` go :)

hebasto commented:
@practicalswift 
> Great to see `deadlock:UpdateTip` go :)

I must confess that, unlike #19982, I'm not as mush confident in this PR removals as all of them are found empirically.

hebasto commented:
Some more CI TSan builds:
- https://cirrus-ci.com/task/4866563041918976
- https://cirrus-ci.com/task/5574751877857280

hebasto commented:
Rebased 49c3a5287d74d3739724d9f7f6c7cff3ec398330 -> 730e5ffc2e9f86fded9283543c0e8dc0759d6c6f ([pr19983.02](https://github.com/hebasto/bitcoin/commits/pr19983.02) -> [pr19983.03](https://github.com/hebasto/bitcoin/commits/pr19983.03)) due to the conflict with #20285.

hebasto commented:
Rebased 71518e39d8da6b55e0ad04acc0b6781290573425 -> cf1ac36915d1f7898a609e4a2c02f1a800953984 ([pr19983.04](https://github.com/hebasto/bitcoin/commits/pr19983.04) -> [pr19983.05](https://github.com/hebasto/bitcoin/commits/pr19983.05)) due to the conflict with #19337.

hebasto commented:
Rebased 71518e39d8da6b55e0ad04acc0b6781290573425 -> 27752af19a701702e94fbf721a4ba2d2c0a7d0ea ([pr19983.04](https://github.com/hebasto/bitcoin/commits/pr19983.04) -> [pr19983.06](https://github.com/hebasto/bitcoin/commits/pr19983.06)) due to the conflict with #19337.

hebasto commented:
Rebased 3c2e969172cbb2796441eb218e5102a19a1bfd5d -> 89952ae75cc1f0c2e7ec225bbef074f411e3e327 ([pr19983.07](https://github.com/hebasto/bitcoin/commits/pr19983.07) -> [pr19983.08](https://github.com/hebasto/bitcoin/commits/pr19983.08)) due to the conflict with #19425.

hebasto commented:
Rebased 89952ae75cc1f0c2e7ec225bbef074f411e3e327 -> b0b6092874f717f3156c043e803600938c7ff571 ([pr19983.08](https://github.com/hebasto/bitcoin/commits/pr19983.08) -> [pr19983.09](https://github.com/hebasto/bitcoin/commits/pr19983.09)) due to the conflict with #19982.

practicalswift commented:
ACK b0b6092874f717f3156c043e803600938c7ff571: patch looks correct and doesn't touch any non-test code

hebasto commented:
Rebased b0b6092874f717f3156c043e803600938c7ff571 -> a17f69f57e42e08ca95469a9997ec1f49fe2bcf9 ([pr19983.09](https://github.com/hebasto/bitcoin/commits/pr19983.09) -> [pr19983.10](https://github.com/hebasto/bitcoin/commits/pr19983.10)) due to the conflict with #20683.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -308,7 +308,7 @@ def get_wallet_rpc(self, wallet_name):
         return self.version is None or self.version >= ver
 
-    def stop_node(self, expected_stderr='', *, wait=0, wait_until_stopped=True):
+    def stop_node(self, expected_stderr='', ignored_stderr='', *, wait=0, wait_until_stopped=True):
```
```suggestion
    def stop_node(self, expected_stderr='', *, ignored_stderr='', wait=0, wait_until_stopped=True):
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19983#issuecomment-749477119).

hebasto commented:
Updated a17f69f57e42e08ca95469a9997ec1f49fe2bcf9 -> 354d2cba44747ca34b6833e0f58901394a03ae49 ([pr19983.10](https://github.com/hebasto/bitcoin/commits/pr19983.10) -> [pr19983.11](https://github.com/hebasto/bitcoin/commits/pr19983.11), [diff](https://github.com/hebasto/bitcoin/compare/pr19983.10..pr19983.11)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19983#discussion_r547181351)

practicalswift commented:
cr ACK 354d2cba44747ca34b6833e0f58901394a03ae49: patch looks correct and doesn't touch any non-test code

MarcoFalke commented on test/functional/feature_maxuploadtarget.py:
```diff
@@ -161,6 +161,9 @@ def run_test(self):
         assert_equal(peer_info[0]['permissions'], ['download'])
 
+        # Ignore non-critical TSan warnings.
+        self.stop_node(0, ignored_stderr="WARNING: too long mutex cycle found")
```
I can't reproduce this locally. How often did you have to run this test?

I am running with:

```diff
diff --git a/test/sanitizer_suppressions/tsan b/test/sanitizer_suppressions/tsan
index 986e09605..020f3da9b 100644
--- a/test/sanitizer_suppressions/tsan
+++ b/test/sanitizer_suppressions/tsan
@@ -29,12 +29,7 @@ race:zmq::*
 race:bitcoin-qt
 # deadlock (TODO fix)
 deadlock:CConnman::ForNode
-deadlock:CConnman::GetNodeStats
 deadlock:CChainState::ConnectTip
-deadlock:UpdateTip
-
-# WalletBatch (unidentified deadlock)
-deadlock:WalletBatch
 
 # Intentional deadlock in tests
 deadlock:TestPotentialDeadLockDetected

hebasto replied:
> I can't reproduce this locally.

What is `-stdlib` compiled against?

MarcoFalke replied:
I used `./ci/test/00_setup_env_native_tsan.sh` to compile

hebasto replied:
Indeed. On master (df127ecede804ecb0a70bf45cbca73af3c3e7761) with [your](https://github.com/bitcoin/bitcoin/pull/19983#discussion_r547291853) patch I have no "WARNING: too long mutex cycle found" from the thread sanitizer when running `test/functional/feature_maxuploadtarget.py `. Maybe the reason is the updated clang?

Instead, the skipping `-DDEBUG_LOCKORDER` from `CPPFLAGS` causes another warning even on master w/o your patch:
```
$ test/functional/feature_maxuploadtarget.py 
2020-12-22T15:02:39.792000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_731auno_
2020-12-22T15:02:57.821000Z TestFramework (INFO): Peer 0 disconnected after downloading old block too many times
2020-12-22T15:03:42.603000Z TestFramework (INFO): Peer 1 able to repeatedly download new block
2020-12-22T15:03:42.712000Z TestFramework (INFO): Peer 1 disconnected after trying to download old block
2020-12-22T15:03:42.712000Z TestFramework (INFO): Advancing system time on node to clear counters...
2020-12-22T15:03:42.867000Z TestFramework (INFO): Peer 2 able to download old block
2020-12-22T15:03:42.873000Z TestFramework (INFO): Restarting node 0 with download permission and 1MB maxuploadtarget
2020-12-22T15:03:45.050000Z TestFramework (INFO): Peer still connected after trying to download old block (download permission)
2020-12-22T15:03:45.103000Z TestFramework (INFO): Stopping nodes
Traceback (most recent call last):
  File "test/functional/feature_maxuploadtarget.py", line 166, in <module>
    MaxUploadTest().main()
  File "/home/hebasto/code/gh/bitcoin/test/functional/test_framework/test_framework.py", line 149, in main
    exit_code = self.shutdown()
  File "/home/hebasto/code/gh/bitcoin/test/functional/test_framework/test_framework.py", line 278, in shutdown
    self.stop_nodes()
  File "/home/hebasto/code/gh/bitcoin/test/functional/test_framework/test_framework.py", line 525, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/hebasto/code/gh/bitcoin/test/functional/test_framework/test_node.py", line 333, in stop_node
    raise AssertionError("Unexpected stderr {} != {}".format(stderr, expected_stderr))
AssertionError: Unexpected stderr ==================
WARNING: ThreadSanitizer: double lock of a mutex (pid=365627)
    #0 pthread_mutex_lock <null> (bitcoind+0xa8cf6)
    #1 std::__1::mutex::lock() <null> (libc++.so.1+0x83505)
    #2 CConnman::ThreadSocketHandler() /home/hebasto/code/gh/bitcoin/src/net.cpp:1542:9 (bitcoind+0x1780d4)
    #3 decltype(*(std::__1::forward<CConnman*&>(fp0)).*fp()) std::__1::__invoke<void (CConnman::*&)(), CConnman*&, void>(void (CConnman::*&)(), CConnman*&) /usr/lib/llvm-10/bin/../include/c++/v1/type_traits:3480:1 (bitcoind+0x1a325a)
    #4 std::__1::__bind_return<void (CConnman::*)(), std::__1::tuple<CConnman*>, std::__1::tuple<>, __is_valid_bind_return<void (CConnman::*)(), std::__1::tuple<CConnman*>, std::__1::tuple<> >::value>::type std::__1::__apply_functor<void (CConnman::*)(), std::__1::tuple<CConnman*>, 0ul, std::__1::tuple<> >(void (CConnman::*&)(), std::__1::tuple<CConnman*>&, std::__1::__tuple_indices<0ul>, std::__1::tuple<>&&) /usr/lib/llvm-10/bin/../include/c++/v1/functional:2770:12 (bitcoind+0x1a325a)
    #5 std::__1::__bind_return<void (CConnman::*)(), std::__1::tuple<CConnman*>, std::__1::tuple<>, __is_valid_bind_return<void (CConnman::*)(), std::__1::tuple<CConnman*>, std::__1::tuple<> >::value>::type std::__1::__bind<void (CConnman::*)(), CConnman*>::operator()<>() /usr/lib/llvm-10/bin/../include/c++/v1/functional:2803:20 (bitcoind+0x1a325a)
    #6 decltype(std::__1::forward<std::__1::__bind<void (CConnman::*)(), CConnman*>&>(fp)()) std::__1::__invoke<std::__1::__bind<void (CConnman::*)(), CConnman*>&>(std::__1::__bind<void (CConnman::*)(), CConnman*>&) /usr/lib/llvm-10/bin/../include/c++/v1/type_traits:3539:1 (bitcoind+0x1a325a)
    #7 void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (CConnman::*)(), CConnman*>&>(std::__1::__bind<void (CConnman::*)(), CConnman*>&) /usr/lib/llvm-10/bin/../include/c++/v1/__functional_base:348:9 (bitcoind+0x1a325a)
    #8 std::__1::__function::__alloc_func<std::__1::__bind<void (CConnman::*)(), CConnman*>, std::__1::allocator<std::__1::__bind<void (CConnman::*)(), CConnman*> >, void ()>::operator()() /usr/lib/llvm-10/bin/../include/c++/v1/functional:1540:16 (bitcoind+0x1a325a)
    #9 std::__1::__function::__func<std::__1::__bind<void (CConnman::*)(), CConnman*>, std::__1::allocator<std::__1::__bind<void (CConnman::*)(), CConnman*> >, void ()>::operator()() /usr/lib/llvm-10/bin/../include/c++/v1/functional:1714:12 (bitcoind+0x1a325a)
    #10 std::__1::__function::__value_func<void ()>::operator()() const /usr/lib/llvm-10/bin/../include/c++/v1/functional:1867:16 (bitcoind+0x148cfe)
    #11 std::__1::function<void ()>::operator()() const /usr/lib/llvm-10/bin/../include/c++/v1/functional:2473:12 (bitcoind+0x148cfe)
    #12 void TraceThread<std::__1::function<void ()> >(char const*, std::__1::function<void ()>) /home/hebasto/code/gh/bitcoin/src/./util/system.h:438:9 (bitcoind+0x148cfe)
    #13 decltype(std::__1::forward<void (*)(char const*, std::__1::function<void ()>)>(fp)(std::__1::forward<char const*>(fp0), std::__1::forward<std::__1::function<void ()> >(fp0))) std::__1::__invoke<void (*)(char const*, std::__1::function<void ()>), char const*, std::__1::function<void ()> >(void (*&&)(char const*, std::__1::function<void ()>), char const*&&, std::__1::function<void ()>&&) /usr/lib/llvm-10/bin/../include/c++/v1/type_traits:3539:1 (bitcoind+0x1a342c)
    #14 void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(char const*, std::__1::function<void ()>), char const*, std::__1::function<void ()>, 2ul, 3ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(char const*, std::__1::function<void ()>), char const*, std::__1::function<void ()> >&, std::__1::__tuple_indices<2ul, 3ul>) /usr/lib/llvm-10/bin/../include/c++/v1/thread:273:5 (bitcoind+0x1a342c)
    #15 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(char const*, std::__1::function<void ()>), char const*, std::__1::function<void ()> > >(void*) /usr/lib/llvm-10/bin/../include/c++/v1/thread:284:5 (bitcoind+0x1a342c)

  Location is heap block of size 328920 at 0x7f3445476000 allocated by main thread:
    #0 operator new(unsigned long) <null> (bitcoind+0x11883b)
    #1 std::__1::__unique_if<CConnman>::__unique_single std::__1::make_unique<CConnman, unsigned long, unsigned long, bool>(unsigned long&&, unsigned long&&, bool&&) /usr/lib/llvm-10/bin/../include/c++/v1/memory:3028:28 (bitcoind+0x134039)
    #2 std::__1::unique_ptr<CConnman, std::__1::default_delete<CConnman> > MakeUnique<CConnman, unsigned long, unsigned long, bool>(unsigned long&&, unsigned long&&, bool&&) /home/hebasto/code/gh/bitcoin/src/./util/memory.h:17:12 (bitcoind+0x134039)
    #3 AppInitMain(util::Ref const&, NodeContext&, interfaces::BlockAndHeaderTipInfo*) /home/hebasto/code/gh/bitcoin/src/init.cpp:1386:20 (bitcoind+0x134039)
    #4 AppInit(int, char**) /home/hebasto/code/gh/bitcoin/src/bitcoind.cpp:133:43 (bitcoind+0x11b773)
    #5 main /home/hebasto/code/gh/bitcoin/src/bitcoind.cpp:161:13 (bitcoind+0x11b773)

  Mutex M132842 (0x7f34454c63f8) created at:
    #0 pthread_mutex_lock <null> (bitcoind+0xa8cf6)
    #1 std::__1::mutex::lock() <null> (libc++.so.1+0x83505)
    #2 AppInitMain(util::Ref const&, NodeContext&, interfaces::BlockAndHeaderTipInfo*) /home/hebasto/code/gh/bitcoin/src/init.cpp:1981:24 (bitcoind+0x13d0a3)
    #3 AppInit(int, char**) /home/hebasto/code/gh/bitcoin/src/bitcoind.cpp:133:43 (bitcoind+0x11b773)
    #4 main /home/hebasto/code/gh/bitcoin/src/bitcoind.cpp:161:13 (bitcoind+0x11b773)

SUMMARY: ThreadSanitizer: double lock of a mutex (/home/hebasto/code/gh/bitcoin/src/bitcoind+0xa8cf6) in pthread_mutex_lock
================== != 
[node 0] Cleaning up leftover process
```

hebasto replied:
Going to update this PR after more tests.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19983#issuecomment-749707004l).

hebasto commented:
Updated 354d2cba44747ca34b6833e0f58901394a03ae49 -> 3e1571285f4a0edf59d51bbdeee028be3038b6dc ([pr19983.11](https://github.com/hebasto/bitcoin/commits/pr19983.11) -> [pr19983.12](https://github.com/hebasto/bitcoin/commits/pr19983.12)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19983#discussion_r547291853)
- rebased on top of the #20745

MarcoFalke commented on test/sanitizer_suppressions/tsan:
```diff
@@ -30,13 +30,7 @@ race:zmq::*
-deadlock:UpdateTip
-
-# WalletBatch (unidentified deadlock)
-deadlock:WalletBatch
```
What are the steps to verify this patch? Note that the deadlocks are intermittent, so there is no way to run the tests and reproduce them or not reproduce them reliably.

hebasto replied:
TSan searches for _potential_ deadlocks, i.e., it verifies consistency of lock ordering in the code, right?

hebasto replied:
https://github.com/google/sanitizers/wiki/ThreadSanitizerDeadlockDetector

MarcoFalke replied:
The lock order graph is generated at runtime and is not deterministic. It depends on the code paths that are executed, and our tests exercise different code paths in each run.

hebasto replied:
> What are the steps to verify this patch?

I don't know (besides tracking code changes since these suppressions were introduced).

Having a smaller set of suppressions is better, though. If this patch appears wrong in the next year, it could be reverted :)

MarcoFalke commented:
>  If this patch appears wrong in the next year, it could be reverted :)

Ok, fair enough. Let's give it a try.
